share/pvss/pvss.go

```diff
@@ -152,39 +166,95 @@ func DecShareBatch(suite Suite, H kyber.Point, X []kyber.Point, sH []kyber.Point
-func VerifyDecShareBatch(suite Suite, G kyber.Point, X []kyber.Point, encShares []*PubVerShare, decShares []*PubVerShare) ([]*PubVerShare, error) {
-	if len(X) != len(encShares) || len(encShares) != len(decShares) {
-		return nil, errorDifferentLengths
+func VerifyDecShareBatch(suite Suite, G kyber.Point, pubRun *PubVSSData, decShares []*PubVerShare, expectedChallenge kyber.Scalar) ([]*PubVerShare, error) {
+	// match index to encShare and decShare
+	indToShare := make(map[int]*PubVerShare, pubRun.N)
+	indToPk := make(map[int]kyber.Point, pubRun.N)
+	for i, el := range pubRun.Shares {
+		indToShare[el.S.I] = el
+		indToPk[el.S.I] = pubRun.X[i]
+	}
+
+	filteredDecShares := make([]*PubVerShare, 0)
+	filteredEncShares := make([]*PubVerShare, 0)
+	filteredPks := make([]kyber.Point, 0)
+	for _, decShare := range decShares {
+		if encShare, ok := indToShare[decShare.S.I]; ok {
+			filteredPks = append(filteredPks, indToPk[decShare.S.I])
+			filteredEncShares = append(filteredEncShares, encShare)
+			filteredDecShares = append(filteredDecShares, decShare)
+		}
 	}
+
+	if len(filteredPks) != len(filteredEncShares) || len(filteredEncShares) != len(filteredDecShares) {
+		return nil, ErrDifferentLengths
+	}
+
 	var D []*PubVerShare // good decrypted shares
-	for i := 0; i < len(X); i++ {
-		if err := VerifyDecShare(suite, G, X[i], encShares[i], decShares[i]); err == nil {
-			D = append(D, decShares[i])
+	for i := 0; i < len(filteredPks); i++ {
+		if err := VerifyDecShare(suite, G, filteredPks[i], filteredEncShares[i], filteredDecShares[i]); err != nil {
+			// Expected behavior is to accept some shares may be invalid and keep going
+			continue
+		}
+		if !pubRun.Shares[i].P.C.Equal(expectedChallenge) {
+			return nil, ErrChallengeVerification
 		}
+		D = append(D, decShares[i])
 	}
 	return D, nil
 }
 
 // RecoverSecret first verifies the given decrypted shares against their
 // decryption consistency proofs and then tries to recover the shared secret.
-func RecoverSecret(suite Suite, G kyber.Point, X []kyber.Point, encShares []*PubVerShare, decShares []*PubVerShare, t int, n int) (kyber.Point, error) {
-	D, err := VerifyDecShareBatch(suite, G, X, encShares, decShares)
+func RecoverSecret(suite Suite, G kyber.Point, pubRun *PubVSSData, decShares []*PubVerShare) (kyber.Point, error) {
+	// Recompute the challenge
+	coms := make([]kyber.Point, pubRun.N)
+	ps := make([]kyber.Point, pubRun.N)
+	vG := make([]kyber.Point, pubRun.N)
+	vH := make([]kyber.Point, pubRun.N)
+
+	for i := 0; i < pubRun.N; i++ {
+		coms[i] = pubRun.Shares[i].Com
+		ps[i] = pubRun.Shares[i].S.V
+		vG[i] = pubRun.Shares[i].P.VG
+		vH[i] = pubRun.Shares[i].P.VH
+	}
+
+	h := suite.Hash()
+	for _, x := range coms {
+		x.MarshalTo(h)
+	}
+	for _, x := range ps {
+		x.MarshalTo(h)
+	}
+	for _, x := range vG {
+		x.MarshalTo(h)
+	}
+	for _, x := range vH {
+		x.MarshalTo(h)
+	}
+	cb := h.Sum(nil)
+
+	expectedChallenge := suite.Scalar().Pick(suite.XOF(cb))
```

Comment:
As mentioned in previous comments, the challenge derivation should be moved to `VerifyEncShareBatch`.